fix(deletion): Fix MonitorCheckIn deletion failures

yuvmen opened the pull request:
Our previous attempt (#103495) at correcting processing timeouts for MonitorCheckIn deletions failed , and introduced integrity errors because MonitorCheckIn cant be bulk deleted without manually removing child dependancies (no CASCADE in db FKs).
Added deletion of MonitorIncident directly before attempting to bulk delete MonitorCheckIns, basically removing the children of MonitorCheckins before deleting them.

yuvmen committed:
fix(deletion): Fix MonitorCheckIn deletion failures

Our previous attmept at correcting processing timeouts for MonitorCheckIn deletions failed, and
introduced integrity errors because MonitorCheckIn cant be bulk deleted without manually removing
child dependancies (no CASCADE in db FKs). Instead we add proper deletion tasks to the different
models.
Note we also use `get_child_relations_bulk` here which is actually an old method but not in use
in our codebase today. Using it though we might be able to speed up querying for child relations
of models with complex relations, when they cant just be outright bulk deleted using BulkModelDeletionTask.

yuvmen commented on src/sentry/deletions/defaults/monitor_environment.py:
```diff
@@ -1,9 +1,4 @@
-    ModelDeletionTask,
-    ModelRelation,
-)
+from sentry.deletions.base import BaseRelation, ModelDeletionTask, ModelRelation
```
looks like this is correct and is actually missing, adding it

wedamija replied:
Yeah, I think this can also be done in the same way as `Monitor`, then the checkins can be bulk deleted

wedamija commented on src/sentry/deletions/defaults/monitor.py:
```diff
@@ -12,8 +7,6 @@ def get_child_relations(self, instance: Monitor) -> list[BaseRelation]:
-            ModelRelation(
-                models.MonitorCheckIn, {"monitor_id": instance.id}, BulkModelDeletionTask
-            ),
+            ModelRelation(models.MonitorCheckIn, {"monitor_id": instance.id}),
```
I think we need to add in `MonitorIncident` here, before `MonitorCheckin`. Then all the `MonitorIncident` rows will be removed first, and we can use the bulk delete on `MonitorCheckin` still.

yuvmen committed:
adding MonitorIncident deletions to MonitorEnvironment deletion task

yuvmen commented on src/sentry/deletions/defaults/monitor_incident.py:
```diff
@@ -0,0 +1,15 @@
+                models.MonitorEnvBrokenDetection,
+                {"monitor_incident_id": instance.id},
+            ),
+        ]
```
this is wrong, the model is registered for bulk deletion model and that will be used to delete it efficiently

wedamija commented on src/sentry/deletions/defaults/monitor_incident.py:
```diff
@@ -0,0 +1,14 @@
+
+        return [
+            ModelRelation(
+                models.MonitorEnvBrokenDetection,
```
Yeah, this seems like the right thing to add here

wedamija reviewed (commented):
I think maybe we should cascade this like

Monitor -> MonitorIncident -> MonitorCheckin rather than Monitor -> MonitorCheckin -> MonitorIncident

If we delete the MonitorIncident rows first, then the checkins won't have to cascade delete to the `MonitorIncident` table and you can use the bulk deletion stuff.

This also helps with the `MonitorEnvironment` deletions, since they won't be referenced by `MonitorIncident` either

yuvmen committed:
refactor to leverage child deletion ordering, delete everyting in monitor and monitor_environment flows

allow MonitorIncident to be deleted not in bulk - there shouldnt be so many to cause problems without it
